src/MySqlEventStore.php

```diff
@@ -552,11 +512,166 @@ public function stopProjection(string $name): void
+        if (null === $filter && $regex) {
+            throw new Exception\InvalidArgumentException('No regex pattern given');
+        }
+
+        if ($regex && false === @preg_match("/$filter/", '')) {
+            throw new Exception\InvalidArgumentException('Invalid regex pattern given');
+        }
+
+        $where = [];
+        $values = [];
+
+        if (null !== $filter && $regex) {
+            $where[] = '`real_stream_name` REGEXP :filter ';
+            $values[':filter'] = $filter;
+        } elseif (null !== $filter && ! $regex) {
+            $where[] = '`real_stream_name` = :filter ';
+            $values[':filter'] = $filter;
+        }
+
+        list($where, $values) = $this->createWhereClauseForMetadata($metadataMatcher, $where, $values);
+
+        $whereCondition = implode(' AND ', $where);
+        if (! empty($whereCondition)) {
+            $whereCondition = 'WHERE ' . $whereCondition;
+        }
+
+        $query = <<<SQL
+SELECT `real_stream_name` FROM $this->eventStreamsTable
+$whereCondition
+ORDER BY `no`
+LIMIT $offset, $limit
+SQL;
+
+        $statement = $this->connection->prepare($query);
+        $statement->setFetchMode(PDO::FETCH_OBJ);
+        $statement->execute($values);
+
+        if ($statement->errorCode() !== '00000') {
+            $errorCode = $statement->errorCode();
+            $errorInfo = $statement->errorInfo()[2];
+
+            throw new RuntimeException(
+                "Error $errorCode. Maybe the event streams table is not setup?\nError-Info: $errorInfo"
+            );
+        }
+
+        $result = $statement->fetchAll();
+
+        $streamNames = [];
+
+        foreach ($result as $streamName) {
+            $streamNames[] = new StreamName($streamName->real_stream_name);
+        }
+
+        return $streamNames;
+    }
+
+    public function fetchCategoryNames(?string $filter, bool $regex, int $limit, int $offset): array
+    {
+        if (null === $filter && $regex) {
+            throw new Exception\InvalidArgumentException('No regex pattern given');
+        }
+
+        if ($regex && false === @preg_match("/$filter/", '')) {
+            throw new Exception\InvalidArgumentException('Invalid regex pattern given');
+        }
+
+        $where = [];
+        $values = [];
+
+        if (null !== $filter && $regex) {
+            $where[] = '`real_stream_name` REGEXP :filter ';
+            $values[':filter'] = $filter . '.*-.+';
+        } elseif (null !== $filter && ! $regex) {
+            $where[] = '`real_stream_name` LIKE :filter ';
+            $values[':filter'] = $filter . '-%';
+        }
+
+        $where[] = 'substring(real_stream_name, 1, locate(\'-\', real_stream_name) - 1) != \'\'';
+
+        $whereCondition = 'WHERE ' . implode(' AND ', $where);
+
+        $query = <<<SQL
+SELECT substring(real_stream_name, 1, locate('-', real_stream_name) - 1) as `category` FROM $this->eventStreamsTable
+$whereCondition
+GROUP BY `category`
+LIMIT $offset, $limit
```

Comment:
If we need no order we shoud add `ORDER BY NULL`. This is a trick to avoid filesystem / temp table sorts. But we should check what MySQL `explain` said.